rfc/egress-billing.md

```diff
@@ -0,0 +1,309 @@
+    case Maybe:
+      return { status: 200, body: readContentFromW3up(cid) }
+  }
+}
+```
+
+### User Configuration
+
+In order to allow our users to use uploaded content in their applications and guarantee acceptable read performance,
+we propose use the "[content commitments]" generated by the w3up [blob upload service].
+
+```js
+{ // Content commitment with CID bafy...site
+  // Storage node authorizes access to the stored resource
+  "iss": "did:web:asia.web3.storage",
+  // To Alice
+  "aud": "did:key:zAlice",
+  // Indefinitely
+  "exp": null,
+  "cmd": "/assert/location",
+  // Subject is storage node because they own that resource
+  "sub": "did:web:asia.web3.storage",
+  "pol": [
+    // multihash must match an uploaded blob
+    ["==", ".content", { "/": { "bytes": "mEi...sfKg" } }],
+    // must be available from this url
+    ["==", ".url", "https://asia.w3s.link/ipfs/bafk...7fi"],
+  ]
+}
+```
+
+[TODO: should I translate this into UCAN.current too?]
+
+The "content commitment" above, with CID `bafy...site`, represents a claim that content will be available from
+the URL `https://asia.w3s.link/ipfs/bafk...7fi` indefinitely. Alice can use it to create and sign the following
+delegation to the w3s.link gateway:
+
+```js
+{ // bafy..auth
+  "iss": "did:key:zAlice",
+  "aud": "did:web:w3s.link",
+  "exp": 1716235987 // restrict to a month
+  "cmd": "/assert/location",
+  // Subject is storage node because they own that resource
+  "sub": "did:web:asia.web3.storage",
+  "pol": [
+    // Request origin header must be example.com
+    ["==", ".headers['origin']", "example.com"],
+    // Request URL must have a query string that includes "token=zrptvx"
+    ["==", ".query.token", "zrptvx"]
+  ],
+  "meta": { "proof": { "/": "bafy...site" } }
+}
+```
+
+This delegation authorizes the gateway to serve this content when the HTTP request's `Origin` header has a
+value of `example.com` and the request URL has a query string that includes `token=zrptvx`. This functions
+as an "unlimited" rate limit and can be used to authorize, for example, unlimited egress billing for requests.
+
+Additional `pol` entries can be added to set more stringent rate limits,
+for example specifying that a particular content commitment may
+only be used 100 times per minute or 1000 total. The syntax and
+specifics of these rate limits will be defined in a formal specification.
+
+```mermaid
+sequenceDiagram
+  Alice->>Blob Service: Please store bafk...7fi
+  Blob Service-->>Alice: Stored - you can present bafy...site to fetch bafk...7fi
+  Alice-)Gateway: You can present bafy...auth to fetch bafk...7fi
+```
+
+#### Rate Limit Accounting
+
+When the gateway receives a request, it will query the [content claims service] for claims about the requested
+CID [TODO: and also other information like origin and token?]. If multiple claims establishing different rate limits
+are found, one is chosen randomly and its request count is incremented, resulting in billing to the 
```

Comment:
There's an important clarification I hope I'm reading right.

When are multiple claims found?

If you have to match token and origin, why would you have multiple claims? 

Or maybe you have the matching unlimited claim and the general open access rate limited claim? In which case I'd lean towards using the unlimited claim rather than picking at random.